syntax = "proto3";

package crud;

// User message represents our entity
message User {
  int32 id = 1;
  string name = 2;
  string email = 3;
  string phone = 4;
}

// Request for creating a user
message CreateUserRequest {
  string name = 1;
  string email = 2;
  string phone = 3;
}

// Response after creating a user
message CreateUserResponse {
  User user = 1;
}

// Request for getting a user by ID
message GetUserRequest {
  int32 id = 1;
}

// Response for getting a user
message GetUserResponse {
  User user = 1;
}

// Request for updating a user
message UpdateUserRequest {
  int32 id = 1;
  string name = 2;
  string email = 3;
  string phone = 4;
}

// Response after updating a user
message UpdateUserResponse {
  User user = 1;
}

// Request for deleting a user
message DeleteUserRequest {
  int32 id = 1;
}

// Response after deleting a user
message DeleteUserResponse {
  bool success = 1;
}

// Request for listing all users
message ListUsersRequest {
  // Optional pagination parameters
  int32 page = 1;
  int32 size = 2;
}

// Response for listing users
message ListUsersResponse {
  repeated User users = 1;
  int32 total = 2;
}

// Define the CRUD service
service CrudService {
  // Create a new user
  rpc CreateUser(CreateUserRequest) returns (CreateUserResponse);
  
  // Get a user by ID
  rpc GetUser(GetUserRequest) returns (GetUserResponse);
  
  // Update an existing user
  rpc UpdateUser(UpdateUserRequest) returns (UpdateUserResponse);
  
  // Delete a user
  rpc DeleteUser(DeleteUserRequest) returns (DeleteUserResponse);
  
  // List all users
  rpc ListUsers(ListUsersRequest) returns (ListUsersResponse);
}